core/src/bytecode/value/mod.rs

```diff
@@ -517,48 +694,93 @@ impl TryFrom<u8> for BodyTag {
     }
 }
 
-/// A one-word header for a heap-allocated Nickel value. The layout is as follows (size is given
-/// in bits):
+/// A 56-bits ref-count represented as a sequence of 7 bytes (little endian).
+#[derive(Clone, Copy, Eq, PartialEq, Debug)]
+pub struct RefCount([u8; 7]);
+
+impl RefCount {
+    /// The maximum representable ref count as a `u64`.
+    pub const MAX: u64 = 0x00_FF_FF_FF_FF_FF_FF_FF;
+    /// A ref count equals to one.
+    pub const ONE: Self = RefCount([1, 0, 0, 0, 0, 0, 0]);
+
+    /// Creates a ref count from a `u64` value. As opposed to the [TryFrom] instance that is
+    /// failing for `value > RefCount::MAX`, this method truncates the higher bits and will return
+    /// `RefCount::MAX` in those cases instead.
+    pub fn from_u64_truncate(value: u64) -> Self {
+        let mut data = [0; 7];
+        let bytes = value.to_le_bytes();
+        data.copy_from_slice(&bytes[0..7]);
+        Self(data)
+    }
+}
+
+impl From<RefCount> for u64 {
+    fn from(value: RefCount) -> Self {
+        let mut data = [0; 8];
+        // Since ref count is in little endian representation, we should copy the first 7 bytes and
+        // keep the last one at zero.
+        data[..7].copy_from_slice(&value.0);
+        u64::from_le_bytes(data)
+    }
+}
+
+impl TryFrom<u64> for RefCount {
+    type Error = ();
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        let mut data = [0; 7];
+        let bytes = value.to_le_bytes();
+
+        if bytes[7] == 0 {
+            data.copy_from_slice(&bytes[0..7]);
+            Ok(RefCount(data))
+        } else {
+            Err(())
+        }
+    }
+}
+
+/// A header for a heap-allocated Nickel value. The layout is as follows (size is given in bits and
+/// is platform-independent):
 ///
 /// ```text
-/// +----------+-------------------------+
-/// | Tag (8)  | (Strong) Ref Count (56) |
-/// +----------+-------------------------+
+/// +----------+-------------------------+----------+
+/// | Tag (8)  | (Strong) Ref Count (56) | Position |
+/// +----------+-------------------------+-----------
```

Comment:
Do we actually rely on this layout? Because the struct is `repr(Rust)`, so I think layout isn't guaranteed.